src/main/cpp/src/decimal_utils.cu

```diff
@@ -675,21 +677,24 @@ struct dec128_multiplier {
 
     chunked256 product = multiply(a, b);
 
-    // Spark does some really odd things that I personally think are a bug
-    // https://issues.apache.org/jira/browse/SPARK-40129
-    // But to match Spark we need to first round the result to a precision of 38
-    // and this is specific to the value in the result of the multiply.
-    // Then we need to round the result to the final scale that we care about.
-    int dec_precision       = precision10(product);
-    int first_div_precision = dec_precision - 38;
+    int dec_precision = precision10(product);
 
     int mult_scale = a_scale + b_scale;
-    if (first_div_precision > 0) {
-      auto const first_div_scale_divisor = pow_ten(first_div_precision).as_128_bits();
-      product                            = divide_and_round(product, first_div_scale_divisor);
 
-      // a_scale and b_scale are negative. first_div_precision is not
-      mult_scale = a_scale + b_scale + first_div_precision;
+    // According to https://issues.apache.org/jira/browse/SPARK-40129
+    // and https://issues.apache.org/jira/browse/SPARK-45786, Spark has a bug in
+    // versions 3.2.4, 3.3.3, 3.4.1, 3.5.0 and 4.0.0 The bug is fixed for later versions but to
+    // match the legacy behavior we need to first round the result to a precision of 38 then we need
+    // to round the result to the final scale that we care about.
+    if (cast_interim_result) {
+      int first_div_precision = dec_precision - 38;
+      if (first_div_precision > 0) {
+        auto const first_div_scale_divisor = pow_ten(first_div_precision).as_128_bits();
+        product                            = divide_and_round(product, first_div_scale_divisor);
+
+        // a_scale and b_scale are negative. first_div_precision is not
+        mult_scale = a_scale + b_scale + first_div_precision;
+      }
```

Comment:
```suggestion
    int const mult_scale = [&]() {
      // According to https://issues.apache.org/jira/browse/SPARK-40129
      // and https://issues.apache.org/jira/browse/SPARK-45786, Spark has a bug in
      // versions 3.2.4, 3.3.3, 3.4.1, 3.5.0 and 4.0.0 The bug is fixed for later versions but to
      // match the legacy behavior we need to first round the result to a precision of 38 then we need
      // to round the result to the final scale that we care about.
      if (cast_interim_result) {
        int first_div_precision = dec_precision - 38;
        if (first_div_precision > 0) {
          auto const first_div_scale_divisor = pow_ten(first_div_precision).as_128_bits();
          product                            = divide_and_round(product, first_div_scale_divisor);

          // a_scale and b_scale are negative. first_div_precision is not
          return a_scale + b_scale + first_div_precision;
      }
      return a_scale + b_scale;
      })();
```
This isn't 100% correct due to the diff not picking up everything, but the general idea here is that we can make this const if we use an immediately evaluated lambda.

Reply:
It's always evaluating `cast_interim_result` to false which always results in test failures.